### DISCUSSION TITLE
Approvals and user workflows

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#8536] [Ideas]
Often registering and updating users is not as simple as creating a user in Keycloak. There are many cases where external systems need to be integrated, there may be various onboarding approval flows, as well as the ability to in general keep external applications own information about users in sync.

This discussion is for now just a placeholder to discuss these aspects, and is not something we are able to prioritise right now, but will be good start a discussion around this and gather relevant use-cases.

**Comment by stianst** [comment]
Previous design proposal around an approvals system:
https://github.com/keycloak/keycloak-community/blob/master/design/approvals-system.md

**Comment by stianst** [comment]
Prototype around a first milestone:
https://github.com/keycloak/keycloak/pull/7466
https://github.com/vmuzikar/keycloak/tree/user-events

**Comment by stianst** [comment]
Very relevant to this will be the plans around Keycloak.X to support external integrations through gRPC, KNative, Kafka, etc.. We're also looking at a first pass in terms of a better and more reliable way to obtain user related events.

**Comment by stianst** [comment]
@velias @vmuzikar FIY

**Comment by vmuzikar** [comment]
A few interesting topics:
* What user actions do we want to approve/audit?
   * Initiated by admin REST API, account REST API, required actions, auth flows (registration), ... ?
* How to implement an action interception in case it needs to be approved?
   * How to persist "proposed changes"?
   * Different approach based on action type?
      * E.g. disable user in case of registration; store new JSON representation somewhere in case of admin REST API?
   * Sync vs async approval?
 * How to apply approved changes in case of async approval?
    * The flow was interrupted by an async approval request, previous context no longer exists.
    * How to store context?
 * Pre-events and post-events
 * Integration with existing User Profile implementation

**Reply by velias** [reply]
I agree with @vmuzikar about topics which have to be defined.
Mainly supported actions where we want approval are important to be defined as they may affect complexity a lot.
For new user registration, it might be relatively easy with only some flag in DB that user is under approval (so can't login yet), data can live in main data structures.
For updates it may be much more complex, data have to live temporarily somewhere, how to handle new update until old one is approved, etc.

I also expect approval is matter of pre-events, and it expects response with approval result so are special a bit.

Post-events are here mainly to inform about change which fully happened (no response expected), we should discuss this as part of whole Eventing mechanism review related to better Keycloak.X support for external integrations through gRPC, KNative, Kafka, etc..

**Reply by rjeeb** [reply]
Hello team

I'm trying to implement an approval system in Keycloak and found this discusstion, we already created our own custom implementation for this feature and it works on Users and Groups for admin APIs only. The solution basically extends some of the resources and change the behavior to create approval requests instead of having the actualy implementation. I will share a detailed document about what we did in our own custom implementation just to be all aligned.

However, I don't think that this implementation is going to work when having this in keycloak itself, so I believe that we need to have more generic way of doing this.

I will try to figure out a solution and I will make a POC of it to see if it is effective and can be used to implement the remaining phases.

I am happy to implement everything and contribute to keycloak with all the changes needed but first help me to have a good design for such problem :)

some thoughts:

> What user actions do we want to approve/audit?

Based on our experience with customers, following actions were requested to have approvals on:

- Users created from the console
- Users updates from the console
- Users deletion
- Users password change
- Roles assigning
- Groups assigning

> Initiated by admin REST API, account REST API, required actions, auth flows (registration), ... ?

Above operations were requested from admin REST API only

> How to implement an action interception in case it needs to be approved?

Our custom solution was to intercept requests in the resource layer and when the request is approved, we will execute the same logic exist in the resource for this specific operation. Of course, this will not work if we need to implement a more generic solution and to be part of keycloak itself.

> How to persist "proposed changes"?

In our custom implementation, we used the resources themselves and have a duplicate table for users, I do like the idea of having defined type of the changes if these changes have long life-cycle, but if you think of approvals, usually they live for short period of time or otherwise they can be marked is invalid and can be done again. So I do believe that JSON representation of the changes is fair enough.

> Different approach based on action type?
> E.g. disable user in case of registration; store new JSON representation somewhere in case of admin REST API?

Approval system should work regardless of the action type as it accepts changes and execute the approval workflow on it, once approved, each action is responsible of handling these changes. Of course we need more detailed design proposal for this (and I'm working on something)

> Sync vs async approval?

Typically, approvals needs two users and might have hours or days to be approved. So I do believe async is the way to go. In our custom implemention, once we recieved a request to create user, we create an approval request and send back redirect response to the approval resource itself

> How to apply approved changes in case of async approval?

As mentioned above, from approval system point of view, the one that ceates the request, is the one that should handle the approval of it. for example, if we receive create request from admin API, then after the changes got approved, then the admin API should handle the approval and continue the process. (again, this needs more details on how this can be implemented)

> The flow was interrupted by an async approval request, previous context no longer exists.

If context is needed for continue the process when changes approved, then somehow this context should be saved along with the approval request itself so when it gets approved, the context can be retrieved and used to continue the process.

> Integration with existing User Profile implementation

The problem is that some of the validations exist outside the user profile, so when the changes get approved, we need to revalidate everything again. If these validations were moved to the user profile, then of course having it in the user profile makes a lot of sense.

cc: @vmuzikar @stianst

**Comment by kkcmadhu** [comment]
any plans to bring this approval flow in near future? we have customers who are looking for such use a  case
